add: Kickoff

mrshu opened the pull request:
* Refactor the way initial translation and rotation constants are
  specified, ensure they are closer to what the current rules say (i.e.
  make it so that the robot is not in the penalty area initially).

* Add the initial positions for blue/yellow kickoff.

* Add a kickoff procedure which puts the robot closer to the center
  point.

* Refactor the goal checking to use kickoff and send an event to the
  reflog.

* Refactor initial position noise to a float, which can then be updated.

* Fixes #15

Signed-off-by: mr.Shu <mr@shu.io>

mrshu committed:
add: Kickoff

* Refactor the way initial translation and rotation constants are
  specified, ensure they are closer to what the current rules say (i.e.
  make it so that the robot is not in the penalty area initially).

* Add the initial positions for blue/yellow kickoff.

* Add a kickoff procedure which puts the robot closer to the center
  point.

* Refactor the goal checking to use kickoff and send an event to the
  reflog.

* Refactor initial position noise to a float, which can then be updated.

* Fixes #15

Signed-off-by: mr.Shu <mr@shu.io>

Adman commented on controllers/rcj_soccer_referee_supervisor/referee/supervisor.py:
```diff
@@ -216,13 +218,17 @@ def reset_ball_position(self):
         self.ball.setVelocity([0, 0, 0, 0, 0, 0])
         self.progress_chck['ball'].reset()
 
+    def _add_initial_position_noise(self, translation):
```
```suggestion
    def _add_initial_position_noise(self, translation: List[float]) -> List[float]:
```

Adman reviewed (changes requested):
Looking good, some minor changes required ;)

mrshu commented on controllers/rcj_soccer_referee_supervisor/referee/consts.py:
```diff
@@ -33,16 +22,42 @@
 NEUTRAL_SPOTS = {
-   CENTER_NS: (0, 0),
-   YELLOW_LEFT_NS: (-0.3, -0.3),
+   CENTER_NS:        (0,    0),
```
Can we force it not to? I realize that it's probably far from standard and very much Perl-ish but I still feel like this is for the best.

Adman replied:
Yes, we can turn off formating for specic lines with `# fmt: off` and `# fmt: on`.

mrshu commented on controllers/rcj_soccer_referee_supervisor/referee/referee.py:
```diff
@@ -67,18 +70,67 @@ def check_progress(self):
-            self.ball_reset_timer = self.post_goal_wait_time
+
+            team_goal = self.team_name_yellow
+            team_kickoff = Team.BLUE
```
This was actually intentional -- I liked the fact that we could in principle change the vlaue of `Team.BLUE` if we needed to. I am happy to do `Team.BLUE.value` though.

mrshu committed:
Update controllers/rcj_soccer_referee_supervisor/referee/referee.py

Co-authored-by: Adrián Matejov <ado.matejov@gmail.com>

mrshu commented on controllers/rcj_soccer_referee_supervisor/referee/consts.py:
```diff
@@ -33,16 +22,42 @@
+YELLOW_KICKOFF_TRANSLATION = [-0.1, 0.03817, 0]
+
+KICKOFF_TRANSLATION = {
+    Team.BLUE: BLUE_KICKOFF_TRANSLATION,
```
That's a good point. I did not even make it an enum at first but it basically jumped back on me :+1: 

I guess `.value` would work here.

mrshu commented on controllers/rcj_soccer_referee_supervisor/referee/referee.py:
```diff
@@ -67,18 +70,67 @@ def check_progress(self):
+                specified, it will be chosen randomly.
+        """
+        if team not in (Team.BLUE, Team.YELLOW, None):
+            raise ValueError(f'Unexpected team name {team}')
```
That's very good feedback, will fix.

mrshu committed:
Update controllers/rcj_soccer_referee_supervisor/referee/supervisor.py

Co-authored-by: Adrián Matejov <ado.matejov@gmail.com>

mrshu commented on controllers/rcj_soccer_referee_supervisor/referee/referee.py:
```diff
@@ -67,18 +70,67 @@ def check_progress(self):
+            raise ValueError(f'Unexpected team name {team}')
+
+        if not team:
+            team = Team.BLUE if random.random() > 0.5 else Team.YELLOW
```
The cool thing here is that we should already know about it, thanks to the fact that we have a JSONL log from the first half.

What I had in mind for the second half was something like the following:
1. Figure out who kicked off the first in the first half
    ```bash
    $ cat The_Blues_vs_The_Yellows-2021-01-02T11:48:24.890235Z.jsonl | grep KICKOFF | head -n 1 | jq .
    "Y3"
    ```
2. Add an env variable (something like `RCJ_SOCCERSIM_KICKOFF_TEAM`) to `B` would be set in case yellow was kicking off in the first half and `Y` otherwise
3. Run the second half similarly to the first one

Adman replied:
Cool, sounds good to me.

mrshu commented on controllers/rcj_soccer_referee_supervisor/referee/referee.py:
```diff
@@ -67,18 +70,67 @@ def check_progress(self):
+            )
+
+            # Let the team that did not score the goal have a kickoff.
+            self.kickoff(team_kickoff)
```
As mentioned in the other comment, we should have that information in the JSON log so believe we ought to be covered.

As it currently looks like, we'll run two separate halves.

mrshu replied:
Sorry @Adman I guess I know what you mean now. There was indeed a bug but the latest commit ought to fix it.

We now remember which team ought to be used for the next kickoff in a specific variable.

mrshu committed:
Update controllers/rcj_soccer_referee_supervisor/referee/supervisor.py

Co-authored-by: Adrián Matejov <ado.matejov@gmail.com>

mrshu committed:
fix: Fix inconsistent quotes in referee.py

* Ensure all quotes are double quotes.

Signed-off-by: mr.Shu <mr@shu.io>

mrshu committed:
refactor: Use str values instead of enums

* Although the code will always refer to them as enums, using `str`s
  where possible seems like a better choice instead of proper enums,
  which are hashable but a bit tougher to work with.

* Fix the situation around "who ought to kick after the time" by
  remembering who was it that was supposed to kick when a goal gets
  scored.

Signed-off-by: mr.Shu <mr@shu.io>

mrshu commented:
@Adman Thank you for your review!

I believe I've addressed the bulk of your points -- please do take another look and let me know what you think.

Thanks!
